core/src/main/java/org/apache/iceberg/util/TableScanUtil.java

```diff
@@ -87,11 +105,44 @@ public static CloseableIterable<CombinedScanTask> planTasks(
                     + file.deletes().stream().mapToLong(ContentFile::fileSizeInBytes).sum(),
                 (1 + file.deletes().size()) * openFileCost);
 
-    return CloseableIterable.transform(
-        CloseableIterable.combine(
-            new BinPacking.PackingIterable<>(splitFiles, splitSize, lookback, weightFunc, true),
-            splitFiles),
-        BaseCombinedScanTask::new);
+    if (preservedPartitionIds != null && !preservedPartitionIds.isEmpty()) {
+      Preconditions.checkArgument(
+          spec != null, "spec can't be null when " + "preservedPartitionIds is not null");
+      StructProjection projectedStruct =
+          StructProjection.create(spec.partitionType(), preservedPartitionIds);
```

Comment:
We should take into account that there may be multiple specs. Also, we need to test both v1 and v2 tables as they have different spec evolution rules.